lib/api/bucketPutCors.js

```diff
@@ -58,30 +59,34 @@ export default function bucketPutCors(authInfo, request, log, callback) {
                     return next(errors.NoSuchBucket);
                 }
                 log.trace('found bucket in metadata');
-                return next(null, bucket, rules);
+                // get corsHeaders before CORSConfiguration is updated
+                const corsHeaders = collectCorsHeaders(request.headers.origin,
+                    request.method, bucket);
+                return next(null, bucket, rules, corsHeaders);
             });
         },
-        function validateBucketAuthorization(bucket, rules, next) {
+        function validateBucketAuthorization(bucket, rules, corsHeaders, next) {
             if (!isBucketAuthorized(bucket, requestType, canonicalID)) {
                 log.debug('access denied for account on bucket', {
                     requestType,
                 });
-                return next(errors.AccessDenied);
+                return next(errors.AccessDenied, corsHeaders);
             }
-            return next(null, bucket, rules);
+            return next(null, bucket, rules, corsHeaders);
         },
-        function updateBucketMetadata(bucket, rules, next) {
+        function updateBucketMetadata(bucket, rules, corsHeaders, next) {
             log.trace('updating bucket cors rules in metadata');
             bucket.setCors(rules);
-            metadata.updateBucket(bucketName, bucket, log, next);
+            metadata.updateBucket(bucketName, bucket, log, err =>
+                next(err, corsHeaders));
         },
-    ], err => {
+    ], (err, corsHeaders) => {
```

Comment:
.. Let me test for a minute because my memory seems different from what I coded / commented.